File: src/ingestion/RowValidator.cpp
#include "ingestion/RowValidator.hpp"
#include "models/Money.hpp"
#include "models/Date.hpp"

namespace reconciliation::ingestion {

using models::Transaction;
using models::Money;
using models::Date;
using models::SourceSystem;

RowValidator::RowValidator(std::unordered_set<std::string> supportedCurrencies)
    : supportedCurrencies_(std::move(supportedCurrencies)) {}

RowValidationResult RowValidator::validateRow(const std::vector<std::string>& fields,
                                               std::size_t rowNumber,
                                               SourceSystem source) const {
    RowValidationResult result;

    if (fields.size() != EXPECTED_FIELD_COUNT) {
        result.errorMessage = "expected " + std::to_string(EXPECTED_FIELD_COUNT) +
                               " columns, got " + std::to_string(fields.size());
        return result;
    }

    const std::string& transactionId = fields[0];
    const std::string& invoiceNumber = fields[1];
    const std::string& vendorId      = fields[2];
    const std::string& dateField     = fields[3];
    const std::string& amountField   = fields[4];
    const std::string& taxField      = fields[5];
    const std::string& currency      = fields[6];

    if (transactionId.empty()) {
        result.errorMessage = "missing transaction_id";
        return result;
    }
    if (invoiceNumber.empty()) {
        result.errorMessage = "missing invoice_number";
        return result;
    }
    if (vendorId.empty()) {
        result.errorMessage = "missing vendor_id";
        return result;
    }

    Date date;
    try {
        date = Date::parse(dateField);
    } catch (const std::invalid_argument& e) {
        result.errorMessage = std::string("invalid transaction_date: ") + e.what();
        return result;
    }

    Money amount;
    try {
        amount = Money::fromDecimalString(amountField);
    } catch (const std::invalid_argument& e) {
        result.errorMessage = std::string("malformed amount: ") + e.what();
        return result;
    }
    if (amount.cents() < 0) {
        // Business rule for this project: a transaction/invoice amount
        // must be non-negative. A real accounts-payable system might
        // legitimately have negative amounts for credit notes/refunds —
        // if this project needed to support that, the rule would move to
        // a per-currency or per-transaction-type policy rather than a
        // blanket rejection. Documented here since it's a deliberate
        // scope decision, not an oversight.
        result.errorMessage = "negative amount not allowed: " + amount.toString();
        return result;
    }

    Money tax;
    try {
        tax = Money::fromDecimalString(taxField);
    } catch (const std::invalid_argument& e) {
        result.errorMessage = std::string("malformed tax_amount: ") + e.what();
        return result;
    }
    if (tax.cents() < 0) {
        result.errorMessage = "negative tax_amount not allowed: " + tax.toString();
        return result;
    }

    if (supportedCurrencies_.find(currency) == supportedCurrencies_.end()) {
        result.errorMessage = "unsupported currency: '" + currency + "'";
        return result;
    }

    Transaction t;
    t.transactionId = transactionId;
    t.invoiceNumber = invoiceNumber;
    t.vendorId = vendorId;
    t.transactionDate = date;
    t.amount = amount;
    t.taxAmount = tax;
    t.currency = currency;
    t.source = source;
    t.sourceRowNumber = rowNumber;

    result.isValid = true;
    result.transaction = std::move(t);
    return result;
}

} // namespace reconciliation::ingestion

File: recon_engine/ingestion/row_validator.py
"""
RowValidator
------------
WHY separate from the ingestor?
    This class touches no files, no database, no batching — it's a pure
    function of (fields in) -> (Transaction or error out). That makes it
    trivial to unit test with hand-written field lists, without needing a
    real CSV file on disk or a database connection.

DESIGN DECISION — duplicate transaction_id is NOT a validation error.
    Duplicates are a real-world data condition the reconciliation engine
    must classify (status DUPLICATE), not data to discard at ingestion.
    Rejecting duplicates here would destroy the exact signal reconciliation
    needs to detect. So RowValidator validates each row independently;
    duplicate detection happens later, during reconciliation itself.

DESIGN DECISION — supported currencies is a fixed, configurable set.
    A real system would likely load this from a currency reference table.
    For this project's scope, a fixed set passed into the constructor is
    sufficient to demonstrate the validation pattern, and keeps this class
    free of a database dependency (it stays pure).
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import FrozenSet, List, Optional

from ..models.date_value import Date, DateParseError
from ..models.money import Money
from ..models.transaction import SourceSystem, Transaction

DEFAULT_CURRENCIES: FrozenSet[str] = frozenset({"INR", "USD", "EUR", "GBP"})
EXPECTED_FIELD_COUNT = 7


@dataclass
class RowValidationResult:
    is_valid: bool = False
    transaction: Optional[Transaction] = None
    error_message: str = ""


class RowValidator:
    def __init__(self, supported_currencies: FrozenSet[str] = DEFAULT_CURRENCIES):
        self._supported_currencies = supported_currencies

    def validate_row(self, fields: List[str], row_number: int, source: SourceSystem) -> RowValidationResult:
        if len(fields) != EXPECTED_FIELD_COUNT:
            return RowValidationResult(
                error_message=f"expected {EXPECTED_FIELD_COUNT} columns, got {len(fields)}"
            )

        transaction_id, invoice_number, vendor_id, date_field, amount_field, tax_field, currency = fields

        if not transaction_id:
            return RowValidationResult(error_message="missing transaction_id")
        if not invoice_number:
            return RowValidationResult(error_message="missing invoice_number")
        if not vendor_id:
            return RowValidationResult(error_message="missing vendor_id")

        try:
            date = Date.parse(date_field)
        except DateParseError as e:
            return RowValidationResult(error_message=f"invalid transaction_date: {e}")

        try:
            amount = Money.from_decimal_string(amount_field)
        except ValueError as e:
            return RowValidationResult(error_message=f"malformed amount: {e}")
        if amount.cents() < 0:
            # Business rule for this project: a transaction/invoice amount
            # must be non-negative. A real accounts-payable system might
            # legitimately have negative amounts for credit notes/refunds —
            # if this project needed to support that, the rule would move
            # to a per-currency or per-transaction-type policy rather than
            # a blanket rejection. Documented as a deliberate scope
            # decision, not an oversight.
            return RowValidationResult(error_message=f"negative amount not allowed: {amount.to_string()}")

        try:
            tax = Money.from_decimal_string(tax_field)
        except ValueError as e:
            return RowValidationResult(error_message=f"malformed tax_amount: {e}")
        if tax.cents() < 0:
            return RowValidationResult(error_message=f"negative tax_amount not allowed: {tax.to_string()}")

        if currency not in self._supported_currencies:
            return RowValidationResult(error_message=f"unsupported currency: '{currency}'")

        transaction = Transaction(
            transaction_id=transaction_id,
            invoice_number=invoice_number,
            vendor_id=vendor_id,
            transaction_date=date,
            amount=amount,
            tax_amount=tax,
            currency=currency,
            source=source,
            source_row_number=row_number,
        )
        return RowValidationResult(is_valid=True, transaction=transaction)
